### ISSUE TITLE
TextBlock TestAlignment="Right" sometimes cuts off text

### ISSUE THREAD

**Issue by darinkes** [AvaloniaUI/Avalonia#13823]
## Describe the bug
I'm in the process of updating AvaloniaUI from 0.10.18 to 11.0.5.
In one view which uses a TextBlock with TextAlignment="Right" sometimes the Text is cut off.
Weird part is, it doesn't happen all the time.
Also disappears if the Text gets updated. 

## To Reproduce

Tough one. Tried a minimal repo, but couldn't trigger it there.

## Expected behavior

Text is always visible

## Screenshots

Add a 1px Pink Border around the TextBlock to make the dimensions visible.

Init State:
![broken state](https://github.com/AvaloniaUI/Avalonia/assets/637743/066a56b1-c773-4f86-9655-04113808cd59)

One Value got changed:
![half broken state](https://github.com/AvaloniaUI/Avalonia/assets/637743/948d7833-d0a8-434f-aef8-81e2e120f106)

Both Value changed:
![working state](https://github.com/AvaloniaUI/Avalonia/assets/637743/e0e977db-0aa2-4f2b-a793-e5ea35c83475)

## Environment 

- OS: checked Windows 10 and Ubuntu 22.04
- Avalonia-Version: 11.0.5

## Additional context

Not sure if its my fault or Avalonia. But it works with 0.10.18.
Please advice what might be the issue, couldn't figure it out

**Comment by timunie** [comment]
looking further it could also be a missing PropertyChanged notify in your ViewModel 🤔

**Comment by darinkes** [comment]
> maybe inspect the visual tree with dev tools F12 and see if it helps to specify widht, other font, Margin, Padding... No idea here without a real clue how to reproduce.

Yeah. Playing with any property triggers something (Measure, Arrange, Render?) and it looks as expected.
Still trying to trigger it an minimal repo, but no luck so far since I have no idea what triggers it, too /o\

> looking further it could also be a missing PropertyChanged notify in your ViewModel 🤔

Hm, unlikely. The Text-Property always contains the expected value. Its just cut-off.
ClipToBounds=false, shows it.

Setter is as easy as it gets.
```
private ulong _downloadAmount;
public ulong DownloadAmount
{
    get => _downloadAmount;
    set => this.RaiseAndSetIfChanged(ref _downloadAmount, value);
}
```
plus an Converter to make it readable for humans.
```
<TextBlock Grid.Row="1" Grid.Column="1" Text="{Binding DownloadAmount, Converter={StaticResource BytesConverter}}"/>
```

![clip-to-bounds](https://github.com/AvaloniaUI/Avalonia/assets/637743/a108c9b0-c092-49d8-8407-5a44c2d6984e)

(TextBlock Background is colored)

It just happens sometimes on Startup, once the values change it fixes itself.

**Comment by darinkes** [comment]
> Make sure properties are only set on the UiThread in case you load them async.

Ack, made sure threading is correct.

Just noticed that while playing with the AvaloniaDebugger.

If I resize the Window, the TextBlock grows with the UI. But the Text does not move.
![no-right-alignment](https://github.com/AvaloniaUI/Avalonia/assets/637743/fa644ba0-c873-4589-a17f-7890c1275c7e)

To the moment the Text changes, then it gets moved to the expected right alignment position.
![with-right-alignment](https://github.com/AvaloniaUI/Avalonia/assets/637743/8334da2c-d9da-413e-9b44-d7bf77cfa96a)

**Comment by timunie** [comment]
okay that seems to be a text layout issue then.

/cc @Gillibald

**Comment by busitech** [comment]
TextPresenter also cuts off the right edge of text when alignment is set to right.

According to the source code (line 630), TextPresenter uses logic copied from TextBlock for measuring the width of text (for example).  We infer that the same problems mentioned in this issue also affect TextPresenter.

We're presently using an Avalonia commit on the master branch from about a week ago.

**Comment by busitech** [comment]
When Left Justified, the measurement cuts off the text, and it is allowed to overprint the boundary.

<img alt="Image" src="https://github.com/user-attachments/assets/767db4c8-41ef-4097-91cb-c34f7db02a4a" />

When Right Justified, the measurement cuts off the text, and it is not allowed to overprint the boundary.

<img alt="Image" src="https://github.com/user-attachments/assets/c8873bbd-b997-4b0c-a055-77cf69ef49f7" />
